File: app/src/main/java/com/example/ui/MainGameContainer.kt
package com.example.ui

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.example.data.GameRepository
import com.example.data.RunResultEntity
import com.example.game.AudioDirector
import com.example.game.GameEngine
import com.example.game.GameScreenState
import kotlinx.coroutines.launch

@Composable
fun MainGameContainer(modifier: Modifier = Modifier) {
    val context = LocalContext.current
    val repository = remember { GameRepository(context) }
    val audioDirector = remember { AudioDirector(context) }
    val engine = remember { GameEngine(audioDirector) }
    val scope = rememberCoroutineScope()

    val topRuns by repository.topRuns.collectAsStateWithLifecycle(initialValue = emptyList())
    val bestDistance by repository.bestDistance.collectAsStateWithLifecycle(initialValue = 0)
    val totalMoney by repository.totalMoney.collectAsStateWithLifecycle(initialValue = 0)
    val achievements by repository.achievements.collectAsStateWithLifecycle(initialValue = emptyList())

    // Monitor Game State Changes (Save Run on Game Over)
    var previousState by remember { mutableStateOf(engine.state) }
    LaunchedEffect(engine.state) {
        if (previousState == GameScreenState.CRASHING && engine.state == GameScreenState.GAME_OVER) {
            scope.launch {
                repository.saveRun(
                    RunResultEntity(
                        distanceMeters = engine.distanceMeters.toInt(),
                        moneyCollected = engine.moneyTaka,
                        maxCombo = engine.maxCombo,
                        zoneReached = engine.currentZone.displayName
                    )
                )
            }
        }
        previousState = engine.state
    }

    Box(modifier = modifier.fillMaxSize()) {
        when (engine.state) {
            GameScreenState.HOME -> {
                HomeScreen(
                    bestDistanceMeters = bestDistance ?: 0,
                    totalMoneyTaka = totalMoney ?: 0,
                    onStartGame = { engine.startNewGame() },
                    onOpenLeaderboard = { engine.goToLeaderboard() },
                    onOpenAchievements = { engine.goToAchievements() },
                    onOpenSettings = { engine.goToSettings() }
                )
            }
            GameScreenState.PLAYING, GameScreenState.CRASHING, GameScreenState.PAUSED -> {
                GameCanvasView(
                    engine = engine,
                    onPauseClicked = {
                        if (engine.state == GameScreenState.PLAYING) engine.pauseGame()
                        else if (engine.state == GameScreenState.PAUSED) engine.resumeGame()
                    }
                )
            }
            GameScreenState.GAME_OVER -> {
                GameOverScreen(
                    distanceMeters = engine.distanceMeters.toInt(),
                    moneyTaka = engine.moneyTaka,
                    maxCombo = engine.maxCombo,
                    zoneName = engine.currentZone.displayName,
                    onPlayAgain = { engine.startNewGame() },
                    onGoHome = { engine.goToHome() }
                )
            }
            GameScreenState.LEADERBOARD -> {
                LeaderboardScreen(
                    topRuns = topRuns,
                    onBackClicked = { engine.goToHome() }
                )
            }
            GameScreenState.ACHIEVEMENTS -> {
                AchievementsScreen(
                    achievements = achievements,
                    onBackClicked = { engine.goToHome() }
                )
            }
            GameScreenState.SETTINGS -> {
                SettingsScreen(
                    bgmEnabled = audioDirector.bgmEnabled,
                    sfxEnabled = audioDirector.sfxEnabled,
                    voiceEnabled = audioDirector.voiceEnabled,
                    useBanglaDigits = engine.useBanglaDigits,
                    showFpsOverlay = engine.showFpsOverlay,
                    onBgmToggled = { audioDirector.bgmEnabled = it },
                    onSfxToggled = { audioDirector.sfxEnabled = it },
                    onVoiceToggled = { audioDirector.voiceEnabled = it },
                    onBanglaDigitsToggled = { engine.useBanglaDigits = it },
                    onFpsOverlayToggled = { engine.showFpsOverlay = it },
                    onBackClicked = { engine.goToHome() }
                )
            }
        }
    }
}

File: flutter_app/lib/ui/main_game_container.dart
import 'package:flutter/material.dart';
import '../data/game_repository.dart';
import '../game/audio_director.dart';
import '../game/game_engine.dart';
import '../game/game_engine.dart';
import '../ui/achievements_screen.dart';
import '../ui/game_canvas.dart';
import '../ui/game_over_screen.dart';
import '../ui/home_screen.dart';
import '../ui/leaderboard_screen.dart';
import '../ui/settings_screen.dart';

class MainGameContainer extends StatefulWidget {
  const MainGameContainer({super.key});

  @override
  State<MainGameContainer> createState() => _MainGameContainerState();
}

class _MainGameContainerState extends State<MainGameContainer> {
  final GameRepository _repository = GameRepository();
  final AudioDirector _audioDirector = AudioDirector();
  final GameEngine _engine = GameEngine(AudioDirector());

  List<dynamic> _topRuns = [];
  int _bestDistance = 0;
  int _totalMoney = 0;
  List<dynamic> _achievements = [];

  @override
  void initState() {
    super.initState();
    _loadData();
  }

  Future<void> _loadData() async {
    await _repository.load();
    setState(() {
      _topRuns = _repository.topRuns;
      _bestDistance = _repository.bestDistance;
      _totalMoney = _repository.totalMoney;
      _achievements = _repository.achievements;
    });
  }

  @override
  Widget build(BuildContext context) {
    return ListenableBuilder(
      listenable: _engine,
      builder: (context, child) {
        return switch (_engine.state) {
          GameScreenState.home => HomeScreen(
              bestDistanceMeters: _bestDistance,
              totalMoneyTaka: _totalMoney,
              onStartGame: _engine.startNewGame,
              onOpenLeaderboard: _engine.goToLeaderboard,
              onOpenAchievements: _engine.goToAchievements,
              onOpenSettings: _engine.goToSettings,
            ),
          GameScreenState.playing || GameScreenState.crashing || GameScreenState.paused => GameCanvasView(
              engine: _engine,
              onPauseClicked: () {
                if (_engine.state == GameScreenState.playing) {
                  _engine.pauseGame();
                } else if (_engine.state == GameScreenState.paused) {
                  _engine.resumeGame();
                }
              },
            ),
          GameScreenState.gameOver => GameOverScreen(
              distanceMeters: _engine.distanceMeters.toInt(),
              moneyTaka: _engine.moneyTaka,
              maxCombo: _engine.maxCombo,
              zoneName: _engine.currentZone.displayName,
              onPlayAgain: _engine.startNewGame,
              onGoHome: _engine.goToHome,
            ),
          GameScreenState.leaderboard => LeaderboardScreen(
              topRuns: _topRuns.cast(),
              onBackClicked: _engine.goToHome,
            ),
          GameScreenState.achievements => AchievementsScreen(
              achievements: _achievements.cast(),
              onBackClicked: _engine.goToHome,
            ),
          GameScreenState.settings => SettingsScreen(
              bgmEnabled: _audioDirector.bgmEnabled,
              sfxEnabled: _audioDirector.sfxEnabled,
              voiceEnabled: _audioDirector.voiceEnabled,
              useBanglaDigits: _engine.useBanglaDigits,
              showFpsOverlay: _engine.showFpsOverlay,
              onBgmToggled: (v) => _audioDirector.bgmEnabled = v,
              onSfxToggled: (v) => _audioDirector.sfxEnabled = v,
              onVoiceToggled: (v) => _audioDirector.voiceEnabled = v,
              onBanglaDigitsToggled: (v) => _engine.useBanglaDigits = v,
              onFpsOverlayToggled: (v) => _engine.showFpsOverlay = v,
              onBackClicked: _engine.goToHome,
            ),
        };
      },
    );
  }
}
